addons/udes_permissions/models/user_template.py

```diff
@@ -0,0 +1,303 @@
+    )
+    family_acl_count = fields.Integer(
+        compute="_compute_family_counts",
+        help="The number of ACLs associated with this template family (including its own groups)",
+    )
+
+    def get_template_groups(self):
+        """
+        Get the usergroups associated with a template family
+        along with the usergroup set under `template_group_id`.
+
+        :return: res.groups(x,)
+        """
+        self.ensure_one()
+        return self.family_group_ids | self.template_group_id
+
+    def get_template_descendants(self):
+        """
+        Get the template along with all of its descendants
+
+        :return: user.template(x,)
+        """
+        self.ensure_one()
+        return self | self.family_descendant_ids
+
+    @api.model
+    def create(self, vals):
+        """Extend create:
+        Create a res.group for the template and link it to template_group_id
+        """
+        template_group = self._create_group_for_template(vals.get("name"))
+        vals["template_group_id"] = template_group.id
+        return super().create(vals)
+
+    def write(self, vals):
+        """Extend write:
+        If we modify `name` - propagate that change to the `template_group_id.name`
+        """
+        res = super().write(vals)
+        if (
+            "name" in vals
+            and vals.get("name")
+            and not self.env.context.get("bypass_name_propagation_to_group")
+        ):
+            self.template_group_id.with_context(
+                dict(bypass_name_propagation_to_template=True)
+            ).name = self._get_user_template_group_name(vals.get("name"))
+        return res
+
+    def unlink(self):
+        """Extend unlink:
+        If we delete the template, delete its respective template_group_id
+        """
+        self.template_group_id.unlink()
+        return super().unlink()
+
+    def _get_user_template_group_name(self, base_name):
+        """
+        Generate the full user template group name, with prefix
+        :param: basename: str() of the template name
+        """
+        return f"{self.TEMPLATE_GROUP_PREFIX}{base_name}"
+
+    def _set_user_template_group_name(self, group_name):
+        """
+        Given a group name which includes the prefix,
+        update the name of the template by removing the prefix from group name,
+        passing additional context to prevent recursion
+        :param: groupname: str() of the group name
+                           i.e return value of _get_user_template_group_name()
+        """
+        if group_name.startswith(self.TEMPLATE_GROUP_PREFIX):
+            self.with_context(dict(bypass_name_propagation_to_group=True)).name = group_name[
+                len(self.TEMPLATE_GROUP_PREFIX) :
+            ]
+        else:
+            # We don't really care if it fails as it is only a name, but log warn it anyway
+            _logger.warning(
```

Comment:
Technical nit: the f-string will always be created, whereas with `logger.warning('foo %s', 'bar')` the interpolation only occurs if we are actually going to emit a log message.  It's probably not much overhead here - and it's unlikely that warning-level messages would be suppressed - but worth being aware.